전남대 BE_강명덕_2주차 과제(1단계)

Profile-exe opened the pull request:
1주차 피드백을 반영하여 상품 관리 코드를 불러오고 피드백 커밋을 진행했습니다.

## 코드를 작성하며 어려웠던 점
- `ProblemDetail`과 전역 예외처리, 커스텀 응답에 대해 많은 글들을 찾아보았는데 정말 다양한 방식이 있어 어떻게 구현할지 고민이 되었습니다. 구글링에 하루를 전부 투자하며 많은 레퍼런스를 참고했는데 학습 방식이 잘못된 것 같습니다... 당장 구현하고 적용해본 다음 학습을 통해 리팩토링을 진행하는 방식으로 바꾸어보려고 합니다.
   [노션에 관련 내용을 기록](https://button-cycle-241.notion.site/Day-7-a1056111ffd34d0e9a3de484c130d4e8?pvs=4)하며 공부해보았는데, 아직 많이 부족하다는 것을 느꼈습니다.

## 구현 내용
- `GlobalExceptionHandler`는 `ResponseEntityExceptionHandler`를 상속시켜 다른 스프링 예외가 발생하더라도 일관적인 응답이 가능하도록 구현했습니다.

- `ProblemDetail`을 상속받아 커스텀 응답을 구현했습니다. `invalidParams`를 두어 `@Valid`로 인한 검증 예외가 던져진 경우 해당 필드와 이에 대응하는 메시지를 응답에 추가하도록 작성했습니다.
- `@JsonInclude(JsonInclude.Include.NON_NULL)` 어노테이션을 통해 일반적인 다른 예외에서는 `invalidParams` 필드가 응답에 포함되지 않도록 했습니다.

- 예시: 입력한 상품의 이름이 15자 초과, 카카오 문구 포함, 허용되지 않은 특수문자가 들어간 경우
  ```json
  {
    "type": "https://gift.com/docs/error#P004",
    "title": "Invalid input value for product",
    "status": 400,
    "detail": "상품 생성에 필요한 값이 누락되었거나 잘못되었습니다.",
    "instance": "/api/products",
    "code": "P004",
    "invalidParams": [
      {
        "field": "name",
        "message": "상품 이름은 1자 이상 15자 이하로 입력해주세요."
      },
      {
        "field": "name",
        "message": "상품 이름에 '카카오' 문구가 포함되어 있습니다. 담당 MD와 협의 후 사용해주세요."
      },
      {
        "field": "name",
        "message": "상품 이름은 한글, 영문, 숫자, 공백, 특수문자 (), [], +, -, &, /, _ 만 입력 가능합니다."
      }
    ]
  }
  ```


## 코드 리뷰 시, 멘토님이 중점적으로 리뷰해줬으면 하는 부분

- `ProductInfo`: 어노테이션에 대한 문자열 리터럴을 상수 변수로 관리하고싶었습니다. 컴파일 타임에 값이 할당되어야해서 열거형 대신 클래스로 관리했습니다. 이와 같은 방식과 클래스 이름이 적절한지 궁금합니다.

- [`ErrorResponse`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/ErrorResponse.html)까지 공부를 해보았는데 개발 중 적용해보았을때 응답에 정말 많은 내용이 포함되는 것을 확인했습니다. 가독성도 떨어지는 것 같고 불필요한 정보들이 많이 포함되는 것 같아서 `ProblemDetail`까지만 적용했습니다. 웹 표준을 지키기 위해 `ErrorResponse`까지 적용하는게 올바른 방식인지 궁금합니다.

- `ProblemDetail`의 `type`의 존재를 알게되며 에러애 대한 설명을 제공해주는 페이지가 필요하다고 느껴졌습니다. 실무에서도 에러 발생 시 `type` 등의 필드로 해당 에러에 대한 구체적인 정보를 확인할 수 있는 문서를 제공해주는지 궁금합니다.

- 커스텀 예외는 해당 도메인의 `exception` 패키지에서 구현했지만, 전역 예외처리 핸들러는 오직 하나만 존재하도록 구현했습니다. 커스텀 예외들은 도메인이 분리되어있으니 관리하기 편해지는 반면에 도메인이 많아지고 예외가 많아질수록 전역 예외처리 핸들러가 너무 무거워지지는 않을까 걱정이 됩니다. `@RestControllerAdvice`에 패키지 또는 클래스 옵션을 적용해 도메인별로 각각 예외 처리 핸들러를 구현해도 괜찮나요?

Profile-exe committed:
chore(build.gradle): validation 의존성 추가

Profile-exe committed:
docs(README.md): Step1 구현할 기능 목록 작성

Profile-exe committed:
feat(ProductErrorCode): 상품 API 예외에 대한 정보를 관리하는 열거형 구현

Profile-exe committed:
feat(CommonErrorCode): 일반적인 예외에 대한 정보를 관리하는 열거형 구현

Profile-exe committed:
feat: 상품 API에 대한 커스텀 예외 구현

Profile-exe committed:
feat(CustomProblemDetail): 응답 body에 사용할 ProblemDetail 구현

기존의 ProblemDetails를 구현한 서브 클래스로 추가 정보 제공을 위한 필드 정의
- `code`: 비즈니스 로직에서 에러를 판별하기 위해 사용되는 에러와 일대일 연관된 문자열 (P001, C001 등)
- `invalidParams`: `@Valid`로 적용한 검증 로직에서 예외가 발생한 경우 대응하는 필드와 설명을 포함해 응답하기 위한 배열

Profile-exe committed:
feat(ValidationError): `@Valid` 에러 발생 시 필드와 이에 대응하는 메시지를 가지는 클래스 구현

- `CustomProblemDetail`의 `invalidParams` 배열의 요소로 사용

Profile-exe committed:
feat(ProductInfo): 상품 API에 대한 메시지 리터럴을 관리하는 클래스 구현

- 패키지를 `message`로 옮김
- 어노테이션의 값으로 사용하기 위해 `enum`으로 변경하지 않았음
  - 컴파일 타임에 값이 주입되어야 하기 때문에 클래스의 `staic final` 변수로 지정

Profile-exe committed:
feat(ProductReqDto): 상품 등록에 대한 검증 추가

- 상품의 이름은 공백을 포함하여 최대 15자까지 입력 가능
- 특수 문자는 (), [], +, -, &, /, _ 만 허용
- "카카오" 문구가 들어간 경우에는 담당 MD와 협의한 경우에만 사용 가능

패키지를 변경하고 `ProductInfo`를 활용해 리터럴을 상수 변수로 관리

Profile-exe committed:
refactor(ProductResDto): 패키지를 `dto` 내부로 옮김

Profile-exe committed:
refactor(ProductRepository): 상품을 못찾을 시 `ProductNotFoundException` 예외를 던지도록 변경

추가로 메서드 체이닝은 개행 적용

Profile-exe committed:
refactor(ProductController): 상품 API에 대한 검증 적용 및 커스텀 예외를 던지는 로직 추가

Profile-exe committed:
refactor(EndToEndTest): 패키지 변경 적용

Profile-exe committed:
feat(GlobalExceptionHandler): 전역 예외처리 핸들러 구현

- `handleMethodArgumentNotValid`: 파라미터 유효성 검사가 실패할 때 발생하는 예외처리 메서드 재정의
  - 기본적으로 `INVALID_INPUT_VALUE`로 처리
  - 도메인에 따라 `ErrorCode`를 다르게 처리한다. 상품의 경우 `INVALID_INPUT_VALUE_PRODUCT`
  - 검증을 통과하지 못한 필드와 이에 대응하는 설명을 응답에 추가

Profile-exe committed:
feat: 관리자 페이지의 예외 발생애 대한 대응 로직 추가

- 에러가 발생한 경우 `alert()`를 통해 사용자에게 알리고 다시 수행하도록 변경

Profile-exe committed:
typo(ProductController): 불필요한 세미콜론 제거

hocaron commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -36,30 +47,34 @@ public ResponseEntity<List<ProductResDto>> getProducts() {
 
+        if (productId == null || productId == 0L) {
+            throw new ProductCreateException(ProductErrorCode.PRODUCT_CREATE_FAILED);
+        }
```
해당 에러는 언제 발생할 수 있을까요?!😀 
productId == null || productId == 0L 조건이 true 가 되는 경우가 궁금합니다!

Profile-exe replied:
잘못 생각했던 것 같습니다. 해당 메서드에서 null 또는 0L을 반환하는지 확인을 안해봤습니다...

현재 [567a4d54](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/567a4d54c27e6f72287c5a1856b5dcbfd08e7803) 커밋으로 해당 변경사항 적용했습니다.

hocaron commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -36,30 +47,34 @@ public ResponseEntity<List<ProductResDto>> getProducts() {
 
         if (NO_OF_ROWS_AFFECTED.equals(noOfRowsAffected)) {     // 수정된 행의 개수 반환 - 1이면 성공, 0이면 실패
             return ResponseEntity.ok(ProductInfo.PRODUCT_UPDATE_SUCCESS);
         }
```
만약 클라이언트가 정말 없는 데이터를 업데이트하려고 했을 때도 PRODUCT_UPDATE_FAILED 이 발생할 것 같아요! 이때 INTERNAL_SERVER_ERROR 보다는 400대의 에러를 발생시켜주는 것은 어떨까요?! 명덕님 의견이 궁금합니다💭

해당 에러를 반환하는 위치도 어디가 적절한지 고민해보면 좋을 것 같아요😀

Profile-exe replied:
[ee87f9f0](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/ee87f9f03221fdb30df6fbfdc255d56579922c1f) 커밋에서 상품 ID에 해당하는 상품이 존재하는지 `boolean`값을 반환하는 메서드를 구현하고

[567a4d54](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/567a4d54c27e6f72287c5a1856b5dcbfd08e7803) 커밋에서 `validateProductExists` 메서드를 만들어  상품 수정 및 삭제 로직에 최상단에 검증 과정을 추가했습니다.

hocaron commented on src/main/java/gift/product/exception/ProductDeleteException.java:
```diff
@@ -0,0 +1,15 @@
+package gift.product.exception;
+
+public class ProductDeleteException extends RuntimeException {
```
커스텀 예외 처리 아주 좋습니다👍

예외들이 비슷한 필드를 가지거나 핸들러에서 처리하는 방식이 동일한 클래스들이 있는 것 같아요.(ProductCreateException, ProductDeleteException, ProductNotFoundException 등)

분리한 이유가 궁금합니다💭

Profile-exe replied:
예외가 발생한 상황과 예외 이름이 일대일로 대응되어 각각 고유한 식별자를 가져야 한다고 생각했습니다. 

[](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/5d33cb75aefe5ac44a23979272e60d6c471ea5bd) 커밋에서 두가지 작업을 진행했는데요,

- 비즈니스 최상위 예외 구현
- 이를 상속하고 각 예외에 대응하는 에러코드를 넣어 싱글톤으로 구현했습니다.
  - 싱글톤으로 구현한 이유는 여러 사용자들이 해당 예외상황에 봉착한 경우 
  많은 인스턴스가 만들어지기 때문에 어플리케이션 실행 시 한 인스턴스를 만들고 재사용하는 측면이 좋을 것이라 생각했습니다.

hocaron replied:
예외 처리가 단순해져서 가독성이 좋아졌습니다👍👍👍 enum 사용도 좋습니다!

hocaron commented on src/main/java/gift/product/message/ProductInfo.java:
```diff
@@ -0,0 +1,15 @@
+package gift.product.message;
+
+public final class ProductInfo {
+    public static final String PRODUCT_UPDATE_SUCCESS = "상품 수정 성공";
```
ErrorCode 등 재사용 가능하도록 상수화를 시켜놓으셨군요👍

해당 값들을 상수값으로 분리한 이유가 궁금해요💭 
재사용하는 값이 아닌 것 같아 궁금해서 여쭤봅니다😀

Profile-exe replied:
메시지로서 사용되는 문자열 리터럴들을 상수화를 시켜놓으면, 각 메시지들을 편집하고 싶을때 해당 문자열을 IDE에 검색하지 않고도 `ProductInfo`를 찾아서 편집할 수 있어 변경이 용이하다 생각했습니다.

또한 추후 재사용이 가능한 문자열이 추가된다면 이러한 방식이 더욱 장점을 가질 것 같습니다.

hocaron replied:
엔드포인트 응답값을 확인하기 위해 Cotroller 클래스부터 찾을 것 같은데, 응답값을 확인하기 위해 코드를 한번더 타고 들어가야할 수도 있을 것 같아요🤔 재사용하여 한번의 수정으로 모든 부분에 수정을 할 수 있는 것은 장점이지만, 엔드포인트 또는 validation 메세지가 재사용이 되는 부분이 나올까에 대한 고민도 있어요😎

이건 저의 생각이라서 추후에 기능 구현하면서 고민해봐도 좋을 것 같아요!

Profile-exe replied:
단 한번만 사용되며 재사용될 여지가 없는 리터럴들은 일단 문자열로 놔두고 추후 재사용이 확인되면 상수변수로 변환을 고려해봐야겠네요. 답변 감사드립니다! 추후 해당 부분 고려해보겠습니다!

hocaron commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -74,6 +80,8 @@ public Integer updateProductById(Long productId, ProductReqDto productReqDto) {
     }
 
     public Integer deleteProductById(Long productId) {
+        findProductById(productId); // 상품이 없을 경우 예외 발생
```
예외가 발생한다는 것을 메서드 이름에 알 수 있게 만들어볼 수도 있을 것 같아요.

```java
findProductByIdOrThrow
```

명덕님 생각은 어떤지 궁금해요!

Profile-exe replied:
예외가 발생함을 명시하는 것은 좋다고 생각합니다!
다른 함수에서도 예외를 던지긴 하지만 `orThrow()`는 `optional()`을 사용한다는 것을 암묵적으로 나타내는 것 같아서 `optional()`을 사용하는 부분에서 해당 식별자 변경 했습니다.

[e6a7723d](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/e6a7723d09ed0281af64cd9b5770ef6aadba4849)

hocaron commented on src/main/java/gift/product/dto/ProductReqDto.java:
```diff
@@ -0,0 +1,29 @@
+
+public record ProductReqDto(
+
+        // 상품의 이름은 공백을 포함하여 최대 15자까지 입력 가능
+        // 특수 문자는 (), [], +, -, &, /, _ 만 허용
+        // "카카오" 문구가 들어간 경우에는 담당 MD와 협의한 경우에만 사용 가능
```
같이 협업하는 입장에서 코드를 이해하기에 적절한 주석이네요😎

hocaron commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,88 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(ProductNotFoundException.class)
+    public ResponseEntity<CustomProblemDetail> handleProductNotFoundException(ProductNotFoundException ex, WebRequest request) {
+        ErrorCode errorCode = ex.getErrorCode();
+
+        return ResponseEntity.status(errorCode.getHttpStatus())
+                .body(new CustomProblemDetail(errorCode, request));
+    }
+
+    @ExceptionHandler(ProductCreateException.class)
+    public ResponseEntity<CustomProblemDetail> handleProductCreateException(ProductCreateException ex, WebRequest request) {
+        ErrorCode errorCode = ex.getErrorCode();
+
+        return ResponseEntity.status(errorCode.getHttpStatus())
+                .body(new CustomProblemDetail(errorCode, request));
+    }
+
+    @ExceptionHandler(ProductUpdateException.class)
+    public ResponseEntity<CustomProblemDetail> handleProductUpdateException(ProductUpdateException ex, WebRequest request) {
+        ErrorCode errorCode = ex.getErrorCode();
+
+        return ResponseEntity.status(errorCode.getHttpStatus())
+                .body(new CustomProblemDetail(errorCode, request));
+    }
+
+    @ExceptionHandler(ProductDeleteException.class)
+    public ResponseEntity<CustomProblemDetail> handleProductDeleteException(ProductDeleteException ex, WebRequest request) {
+        ErrorCode errorCode = ex.getErrorCode();
+
+        return ResponseEntity.status(errorCode.getHttpStatus())
+                .body(new CustomProblemDetail(errorCode, request));
+    }
```
내부 로직이 동일하여, 예외를 조금 단순화시켜도 좋을 것 같아요😎

```java
    @ExceptionHandler(BusinassException.class)
    public ResponseEntity<CustomProblemDetail> handleProductUpdateException(ProductUpdateException ex, WebRequest request) {
        ErrorCode errorCode = ex.getErrorCode();

        return ResponseEntity.status(errorCode.getHttpStatus())
                .body(new CustomProblemDetail(errorCode, request));
    }
```

Profile-exe replied:
이 피드백 보고 아차 싶었습니다. 덕분에 함수를 좀 더 간결하게 만들 수 있었습니다. 감사합니다!
[fe34f862](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/fe34f8629efe50814a0d3456ce9d56cf1bbbe314) 커밋에 해당 내용 반영했습니다. ㅎㅎ

hocaron commented on src/main/java/gift/product/exception/ProductErrorCode.java:
```diff
@@ -0,0 +1,47 @@
+    PRODUCT_DELETE_FAILED("P007", "Product delete failed", HttpStatus.INTERNAL_SERVER_ERROR, "상품 삭제에 실패했습니다.")
+    ;
+
+    private final String code;
+    private final String title;
+    private final HttpStatus httpStatus;
+    private final String detail;
```
title 은 언제 사용할 수 있을까요?💭 이 필드는 언제 쓰일지 궁금해요!

Profile-exe replied:
[`ProblemDetail`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/ProblemDetail.html)을 상속하면서 해당 필드를 채워줘야하기에 존재했던 필드입니다.

불필요한 필드가 많은 것 같아 상속을 없앴고 `code`, `status`, `message`만 남겼습니다. 응답이 보기 편해졌습니다.

해당 커밋: [4defa88b](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/4defa88b74c9bd29bb13fb3b6199e488c501e693)

- 개선된 예외처리 응답
  ```json
  {
    "code": "P004",
    "status": 400,
    "message": "상품 생성에 필요한 값이 누락되었거나 잘못되었습니다.",
    "invalidParams": [
      {
        "field": "name",
        "message": "상품 이름에 '카카오' 문구가 포함되어 있습니다. 담당 MD와 협의 후 사용해주세요."
      },
      {
        "field": "name",
        "message": "상품 이름은 1자 이상 15자 이하로 입력해주세요."
      }
    ]
  }
  ```

hocaron reviewed (changes requested):
명덕님, 안녕하세요! 2주차를 함께 진행할 호선우 멘토입니다. 이번 주차에서 클래스를 아주 꼼꼼히 분리해주셨군요👍

노션에 꼼꼼히 기록하시면서 공부하시는 것은 너무 좋은 습관이라고 생각합니다💡 저도 도움 받게 되네요! 감사합니다.

같이 작업하는 사람으로서 궁금한 부분이 생겨 코멘트를 남겼습니다. 이번주도 화이팅입니다💪💪💪

1. ProductInfo: 어노테이션에 대한 문자열 리터럴을 상수 변수로 관리하고싶었습니다. 컴파일 타임에 값이 할당되어야해서 열거형 대신 클래스로 관리했습니다. 이와 같은 방식과 클래스 이름이 적절한지 궁금합니다.<br />
좋은 고민입니다. ProductInfo 값이 재사용되지 않는 값인 것 같아요. 응답 메세지 이거나 validation 예외 발생시에 필요한 메시지인데, 분리하신 이유를 들어볼 수 있을까요?!

2. [ErrorResponse](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/ErrorResponse.html)까지 공부를 해보았는데 개발 중 적용해보았을때 응답에 정말 많은 내용이 포함되는 것을 확인했습니다. 가독성도 떨어지는 것 같고 불필요한 정보들이 많이 포함되는 것 같아서 ProblemDetail까지만 적용했습니다. 웹 표준을 지키기 위해 ErrorResponse까지 적용하는게 올바른 방식인지 궁금합니다.<br />
회사마다 다르겠지만, 저희는 아래 3개의 필드를 가진 객체를 사용합니다. 저는 클라이언트에서 에러를 받았을 때 분기처리하기 위한 status 와 메시지를 화면에 뿌려주기 위한 message 만 있으면 된다고 생각해서 에러 필드를 많이 가져가지 않는 편입니다. 
```
    private final String errorCode;
    private final int status;
    private final String message;
```
명덕님 생각은 어떤신가요?!

3. ProblemDetail의 type의 존재를 알게되며 에러애 대한 설명을 제공해주는 페이지가 필요하다고 느껴졌습니다. 실무에서도 에러 발생 시 type 등의 필드로 해당 에러에 대한 구체적인 정보를 확인할 수 있는 문서를 제공해주는지 궁금합니다.<br />
좋은 궁금증입니다. 스웨거나 REST docs 를 사용하여 API 문서를 정의하는데, 각 API 에서 발생할 수 있는 예외 케이스를 함께 제공해드리니다.

5. 커스텀 예외는 해당 도메인의 exception 패키지에서 구현했지만, 전역 예외처리 핸들러는 오직 하나만 존재하도록 구현했습니다. 커스텀 예외들은 도메인이 분리되어있으니 관리하기 편해지는 반면에 도메인이 많아지고 예외가 많아질수록 전역 예외처리 핸들러가 너무 무거워지지는 않을까 걱정이 됩니다. @RestControllerAdvice에 패키지 또는 클래스 옵션을 적용해 도메인별로 각각 예외 처리 핸들러를 구현해도 괜찮나요?<br />
말씀 주신 것처럼 각 장단점이 있을 것 같습니다. 비슷한 예외들이 현재 있는 것 같아서 해당 예외를 하나로 관리한다면 관리 포인트가 줄 것 같다는 생각도 있는데, 어떻게 생각하시나요?!
fy; [관련 코멘트](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130#discussion_r1664268571)

hocaron commented:
이건 제가 회사에서 리뷰 받은 후에 적용하는 방식이라 반영하지 않으셔도 됩니다😀

코멘트에 대한 답변에 b1476db 수정된(리뷰 반영된) 커밋 코드를 함께 적으면 리뷰하는 입장에서 편하게 느끼시더라고요! 공유드려봅니당.

Profile-exe committed:
feat(BusinessException): 비즈니스 로직의 최상위 예외 객체 구현 및 상속

- 비즈니스 로직 관련 예외들은 `BusinessException`을 상속해 구현한다.
- 싱글톤으로 구현하여 어플리케이션 시작 시 단 하나의 인스턴스만 존재하도록 구현.

Profile-exe committed:
refactor(ErrorCode): title, detail을 message로 통합

중복되는 내용에 대해 간단한 응답을 제공하도록 title, detail을 message로 통합

Profile-exe committed:
refactor(ErrorResponse): ProblemDetail 상속을 없애고 ErrorCode에 대응되는 응답으로 구현

- `ProblemDetail`을 상속하면서 발생하는 `type`, `title`, `detail`, `instance` 필드를 반환하지 않도록 상속 제거
- 변경된 `ErrorCode`에 맞추어 `code`, `status`, `message`만 반환하도록 변경

Profile-exe committed:
refactor(GlobalExceptionHandler): 비즈니스 예외 단일 메서드로 핸들링 및 ErrorResponse 변경사항 반영

- 비즈니스 메서드 핸들링은 공통된 처리를 진행하므로 최상위 예외 객체를 `@ExceptionHandler`에 등록해 하나의 메서드로 처리하도록 변경
- `ErrorResponse`는 더이상 `ProblemDetail`을 상속하지 않으므로 `ErrorCode`만 생성자에 넣어주도록 변경

Profile-exe committed:
refactor(ProductRepository): 상품 추가, 수정, 삭제에 대한 체크예외를 받아 비즈니스 예외를 던지도록 변경

Profile-exe committed:
feat(ProductRepository): 상품 ID에 해당하는 상품이 존재하는지 확인하는 메서드 구현

Profile-exe committed:
refactor(ProductController): 불필요한 분기 제거 및 상품 존재 검증메서드 적용

Profile-exe committed:
refactor(ProductController): 사용하지 않는 변수 제거

Profile-exe committed:
refactor(ProductReqDto): 커스텀 어노테이션으로 변경 및 조합해 단일 어노테이션으로 적용

Profile-exe committed:
refactor: optional()로 orThrow()하는 메서드 식별자에 명시

Profile-exe commented:
1번 질문에 대해 답변해주신 내용

>    좋은 고민입니다. ProductInfo 값이 재사용되지 않는 값인 것 같아요. 응답 메세지 이거나 validation 예외 발생시에 필요한 메시지인데, 분리하신 이유를 들어볼 수 있을까요?!

 각 메시지들을 편집하고 싶을때 해당 문자열을 IDE에 검색하지 않고도 ProductInfo를 찾아서 편집할 수 있어 변경이 용이하다 생각했습니다.
[해당 코멘트](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130#discussion_r1665447415)에 답변 올렸습니다!

___

2번 질문에 대해 답변해주신 내용

>    회사마다 다르겠지만, 저희는 아래 3개의 필드를 가진 객체를 사용합니다. 저는 클라이언트에서 에러를 받았을 때 분기처리하기 위한 status 와 메시지를 화면에 뿌려주기 위한 message 만 있으면 된다고 생각해서 에러 필드를 많이 가져가지 않는 편입니다.
> 
> ```
>     private final String errorCode;
>     private final int status;
>     private final String message;
> ```
> 
> 명덕님 생각은 어떤신가요?!

웹 표준이라지만 이렇게 많은 정보를 제공해야하나 고민이 많았습니다. 그래서 구글링과 주변 학우분들의 프로젝트 코드를 살펴보았는데 보편적으로 `code`, `status`, `message`만 예외처리 응답으로 반환하는 것을 확인했습니다. 

그래서 저도 해당 커밋: [4defa88b](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/4defa88b74c9bd29bb13fb3b6199e488c501e693) 에서  `ProblemDetail`을 상속하지 않고 식별자를 `ErrorResponse`로 바꾸어 해당 필드만 두었습니다.

___

4번 질문에 대해 답변해주신 내용

>    말씀 주신 것처럼 각 장단점이 있을 것 같습니다. 비슷한 예외들이 현재 있는 것 같아서 해당 예외를 하나로 관리한다면 관리 포인트가 줄 것 같다는 생각도 있는데, 어떻게 생각하시나요?!
>    fy; [관련 코멘트](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130#discussion_r1664268571)

이부분을 미처 생각하지 못했는데 소중한 피드백 감사드립니다! 덕분에 코드를 깔끔하게 변경할 수 있었습니다.
[fe34f862](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/130/commits/fe34f8629efe50814a0d3456ce9d56cf1bbbe314) 커밋에 해당 내용 반영했습니다!

hocaron reviewed (approved):
명덕님 리뷰 반영해주시느라 고생 많으셨습니다🌟

에러 코드 처리하는 부분이 정말 깔끔합니다👍
다음 단계에서 뵙기 위해 해당 PR 은 머지하겠습니다.
